egs/librispeech/ASR/pruned_transducer_mmi/model.py

```diff
@@ -170,25 +463,117 @@ def forward(
+            am=self.predictor_joiner.encoder_proj(encoder_out),
+            lm=self.predictor_joiner.decoder_proj(predictor_decoder_out),
             ranges=ranges,
         )
 
-        # logits : [B, T, prune_range, vocab_size]
+        # predictor_logits : [B, T, prune_range, vocab_size]
 
         # project_input=False since we applied the decoder's input projections
         # prior to do_rnnt_pruning (this is an optimization for speed).
-        logits = self.joiner(am_pruned, lm_pruned, project_input=False)
+        predictor_logits = self.predictor_joiner(
+            am_pruned, lm_pruned, project_input=False
+        )
 
         with torch.cuda.amp.autocast(enabled=False):
-            pruned_loss = k2.rnnt_loss_pruned(
-                logits=logits.float(),
+            predictor_pruned_loss = k2.rnnt_loss_pruned(
+                logits=predictor_logits.float(),
                 symbols=y_padded,
                 ranges=ranges,
                 termination_symbol=blank_id,
                 boundary=boundary,
                 reduction="sum",
             )
 
-        return (simple_loss, pruned_loss)
+        # pruned rnnt loss for hybrid head
+        hybrid_lm = self.simple_hybrid_lm_proj(hybrid_decoder_out)
+        with torch.cuda.amp.autocast(enabled=False):
+            hybrid_simple_loss, (px_grad, py_grad) = k2.rnnt_loss_simple(
+                lm=hybrid_lm.float(),
+                am=am.float(),
+                symbols=y_padded,
+                termination_symbol=blank_id,
+                external_lm=None,  # external_lm_out_proj,
+                boundary=boundary,
+                modified=True,
+                reduction="sum",
+                return_grad=True,
+            )
+
+        # ranges : [B, T, prune_range]
+        ranges = k2.get_rnnt_prune_ranges(
+            px_grad=px_grad,
+            py_grad=py_grad,
+            boundary=boundary,
+            s_range=prune_range,
+        )
+
+        # am_pruned : [B, T, prune_range, encoder_dim]
+        # lm_pruned : [B, T, prune_range, decoder_dim]
+        am_pruned, lm_pruned = k2.do_rnnt_pruning(
+            am=self.hybrid_joiner.encoder_proj(encoder_out),
+            lm=self.hybrid_joiner.decoder_proj(hybrid_decoder_out),
+            ranges=ranges,
+        )
+
+        # predictor_logits : [B, T, prune_range, vocab_size]
+
+        # project_input=False since we applied the decoder's input projections
+        # prior to do_rnnt_pruning (this is an optimization for speed).
+        hybrid_logits = self.hybrid_joiner(
+            am_pruned, lm_pruned, project_input=False
+        )
+
+        with torch.cuda.amp.autocast(enabled=False):
+            hybrid_pruned_loss = k2.rnnt_loss_pruned(
+                logits=hybrid_logits.float(),
+                symbols=y_padded,
+                ranges=ranges,
+                termination_symbol=blank_id,
+                boundary=boundary,
+                external_lm=external_lm_out_proj,
```

Comment:
The neural lm scores were added to `num_loss` here, it will add to `px` matrix after a gather operation.